File: 백준/Gold/7576. 토마토/토마토.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.*;

public class Main {
    static int graph[][];
    static int n;
    static int m;
    static int dirX[] = {1, -1, 0, 0};
    static int dirY[] = {0, 0, 1, -1};

    static Queue<int[]> q = new LinkedList<>();
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringTokenizer st = new StringTokenizer(br.readLine());

        m = Integer.parseInt(st.nextToken());
        n = Integer.parseInt(st.nextToken());
        graph = new int[n][m];

        int start[] = new int[2];
        for(int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for(int j = 0; j < m; j++){
                int number = Integer.parseInt(st.nextToken());
                graph[i][j] = number;
                if(number == 1) {
                    q.add(new int[] {i,j});
                }
            }
        }

        bw.write(bfs() + "\n");
        bw.flush();
        bw.close();
    }
    private static int bfs() {
        while(!q.isEmpty()) {
            int[] current = q.poll();
            int dx = current[0];
            int dy = current[1];
            for(int i = 0; i < 4; i++) {
                int nx = dx + dirX[i];
                int ny = dy + dirY[i];
                if(nx >= 0 && nx < n && ny >= 0 && ny < m && graph[nx][ny] == 0) {
                    graph[nx][ny] = graph[dx][dy] + 1;
                    q.add(new int[] {nx, ny});
                }
            }
        }
        int max = Integer.MIN_VALUE;
        if(!checkTomato()) {
            return -1;
        } else {
            for(int i = 0; i<n; i++){
                for(int j = 0; j<m; j++) {
                    if(max < graph[i][j]) {
                        max = graph[i][j];
                    }
                }
            }
            return max -1;
        }
        
    }
    static boolean checkTomato() {
        for(int i = 0; i<n ;i++) {
            for(int j =0; j<m; j++) {
                if(graph[i][j] == 0) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: 백준/Gold/7576. 토마토/토마토.py
from collections import deque

m, n = map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(n)]
dx = [1, -1, 0, 0]
dy = [0, 0, -1, 1]

queue = deque()
for i in range(n):
    for j in range(m):
        if graph[i][j] == 1:
            queue.append((i, j))

def bfs():
    while queue:
            x, y = queue.popleft()
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                if 0 <= nx < n and 0 <= ny < m:
                    if graph[nx][ny] == 0:
                        graph[nx][ny] = graph[x][y] + 1
                        queue.append((nx, ny))

bfs()
ans = 0
for i in range(n):
    for j in range(m):
        if graph[i][j] == 0:
            print(-1)
            exit()
        else:
            ans = max(ans, graph[i][j])

print(ans-1)
